olive/evaluator/olive_evaluator.py

```diff
@@ -216,15 +215,18 @@ def evaluate(
 
     @staticmethod
     def generate_metric_user_config_with_model_io(metric: Metric, model: "OliveModelHandler"):
```

Comment:
if we want the previous behavior, I think the logic is: 
- if metric.data_config, return
- if not io_config, return
- if not io_config.input_shapes, return
- if not is_io_config_static, return 
- create new dummy config from io_config

If the metric comes with its own data config, then we should not do anything with it even though the type is dummy_dataset.

Reply:
in the perf tuning pass, it can just pass data_config is None if not provided.

Reply:
> If the metric comes with its own data config, then we should not do anything with it even though the type is dummy_dataset.

That would break auto perf tuning. This pass is automatically added to the workflow and so don't come with a data config. In some other cases, the metric defines a dummy dataset but doesn't provide the required input_xxxx fields. I think there a lot of assumptions about how/when/where we get this input if not provided. 

Your proposed logic will break a lot of examples and tests. I don't like the conditional branching either but this is the safe solution to move forward for this PR. I can add another task on my plate to revisit this and remove the branching.

Reply:
> create new dummy config from io_config

This assumes that the logic only runs for the latency metric, which in itself is a bad assumption at this level. At the perf tuning level, it is acceptable but not here.

Reply:
I think my proposed logic is what is currently happening. The priority was dataloader_func > data_config > metric io_config > model_io_config. 

In my proposed logic, (dataloader_func and data_config) is 1, metric io_config is also step 1 with user provided data config, rest of the logic is for model io_config

Reply:
if the perf tuning pass just provides none as data config if not provided, the evaluator would automatically fall back on the model's io_config.

Reply:
> if the perf tuning pass just provides none as data config if not provided, the evaluator would automatically fall back on the model's io_config.

Evaluator never created the `DataConfig` itself. It only populated the config if the expected keys were not found. You are proposing creating `DataConfig` in evaluator with default type being 'DummyDataContainer`.  That is not the same.

Currently, PerfTuning pass sends in a data config with the `dummy_dataset` and the evaluator is only populating the fields. Instead of the PerfTuning it could be any other pass or coming directly from user config as well.

However, in your logic, if the user provide a dummy data config without providing the input_xxx the workflow will fail. Your logic forces the user to repeat what's already in the input model.

Reply:
Evaluator created the data config here https://github.com/microsoft/Olive/blob/fc4fc903f09da82d3205a79a011a7bf89089f955/olive/evaluator/olive_evaluator.py#L309. This case is only true if the conditions I listed above are met (except for the static io condition which is needed but was overlooked)

if the user wants the models io config to be used, they can just choose to not provide a `data_config` to the metric config. That's the behavior currently. they don't provide dataloader_func or data config. 
the perf tuning can do the same

Reply:
https://github.com/microsoft/Olive/blob/fc4fc903f09da82d3205a79a011a7bf89089f955/olive/evaluator/olive_evaluator.py#L194

Reply:
Updated.